post-repo-inmem/src/main/java/klon/post/repo/inmem/PostRepositoryInmemImpl.java

```diff
@@ -0,0 +1,48 @@
+package klon.post.repo.inmem;
+
+import klon.post.repo.api.Post;
+import klon.post.repo.api.PostRepository;
+import lombok.Builder;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+@Builder
+public class PostRepositoryInmemImpl implements PostRepository {
+
+    private final ConcurrentHashMap<String, List<InmemPost>> posts = new ConcurrentHashMap<>();
+
+    private final PostMapper mapper;
+
+    @Override
+    public void addPost(Post post) {
+        InmemPost inmemPost = mapper.toInmemPost(post);
```

Comment:
Wydaje mi się, że to wszystko się rozsypie jeśli chodzi o współbieżność.
Teoretycznie może się zdażyć, że dwa nowe posty tego samego usera wjadą na dwóch osobnych wątkach. W zależności od tego który jako ostatni dojdzie do linii 30, ten wygra, a pozostałe posty znikną.
ConcurrentHashMap jest ok (chociaż nie wykorzystujesz w tym momencie tego _Concurrent_), ale wartości, czyli Listy, też powinny być concurrent. I nie podmiana w HashMapie, a doklejanie do list. Nie wiem w tym momencie która klasa będzie najlepsza dla tej listy.  Problem z pobieraniem być może nulla z HashMapy do załatwienia za pomocą [tego](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-). Słowo klucz - atomically.

Generalnie rozwiązań problemu można sobie wyobrazić kilka i pewnie trzeba wybrać to najprostsze, ponieważ moduły inmem i tak będą do wywalenia, bo się nie nadają do uruchomienia w więcej niż jednej instancji.

Zdecydowanie polecam [tę serię](http://tutorials.jenkov.com/java-concurrency/index.html) odnoście Java Concurency. Inne autor też ma fajne.

Reply:
Racja. Też sugerowałem się tym, że moduły inmem nam się nie przydadzą w przyszłości więc nie przyłożyłem do tego w ogóle uwagi